File: Java/src/com/company/Tree/LC109_sortedListToBST.java
package com.company.Tree;

public class LC109_sortedListToBST {
    public TreeNode sortedListToBST(ListNode head) {
        if(head == null) return null;
        ListNode fast = head;
        ListNode slow = head;
        while(fast.next != null && fast.next.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode temp = slow.next;
        slow.next = null;
        TreeNode root = new TreeNode(slow.val);
        root.left = sortedListToBST(head);
        root.right = sortedListToBST(temp);
        return root;
    }

    public static void main(String[] args) {

        ListNode root = new ListNode(-10);
        root.next = new ListNode(-3);
        root.next.next = new ListNode(0);
        root.next.next.next = new ListNode(5);
        root.next.next.next.next = new ListNode(9);

        LC109_sortedListToBST L = new LC109_sortedListToBST();
        TreeNode res = L.sortedListToBST(root);
        System.out.println(res.val);

    }
}

File: python/Tree/LC109_sortedListtoBST.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    n = None

    def sortedListToBST(self, head: ListNode) -> TreeNode:
         # optimal O(2n) time O(logn) space for recursive tree
        # inorder traverse a bst will get asc sorted list
        global n
        n = head
        size = 0

        while(head):
            head = head.next
            size += 1
        return self.build(0, size-1)

    def build(self, start, end):
        global n
        if start > end:
            return None
        m = start + (end-start)//2
        left = self.build(start, m-1)
        parent_node = TreeNode(n.val)
        n = n.next
        parent_node.left = left
        right = self.build(m+1, end)
        parent_node.right = right
        return parent_node

        # O(2n) time O(n) space
    # def sortedListToBST(self, head: ListNode) -> TreeNode:

#         arr = []
#         while(head):
#             arr.append(head.val)
#             head = head.next
#         return self.build(arr, 0, len(arr)-1)

#     def build(self, arr, l, r):
#         if l>r:
#             return None
#         m = l+(r-l)//2
#         root = TreeNode(arr[m])
#         root.left = self.build(arr, l, m-1)
#         root.right = self.build(arr, m+1, r)
#         return root

        # O(nlogn) time O(logn) space for binary tree
    # def sortedListToBST(self, head: ListNode) -> TreeNode:

#         if not head:
#             return None
#         if not head.next:
#             return TreeNode(head.val)

#         prev = None
#         slow = head
#         fast = head
        # this step cost O(n/2)
#         while(fast and fast.next):
#             prev = slow
#             slow = slow.next
#             fast = fast.next.next
#         root = TreeNode(slow.val)

#         if prev:
#             prev.next = None

#         root.left = self.sortedListToBST(head)
#         root.right = self.sortedListToBST(slow.next)
#         return root
